### PR TITLE
Enhance assign workflow

### LINKED ISSUE #216: Assign - enhanced workflow
1. `suggest DURATION t/TASK`: find the driver who has the earliest time to fit in this task. Fails when the slot cannot be found.

2. `assign t/TASK d/DRIVER at/TIME`: assume that the manager know what he/she is doing; essentially the `assign force` now, but will still show suggested time.

### PR DESCRIPTION
**Pull request by tysg**
Closes #216 
* Introduce `suggest` command, that takes a task and an proposed duration, and automatically suggest an optimal driver and time slot based on:
  - if another driver is delivering to the same customer as this task, try to assign to that driver
  - if the above fails, assign the task to a driver whose has an earliest available time for the task

* Refractor `assign` to just be the old force assign. 
* Refractor `assign force` to override current task allocation

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AssignCommand.java [hunk 1]
```diff
@@ -62,31 +62,80 @@ public AssignCommand(int driverId, int taskId, EventTime eventTime, boolean isFo
...
-        if (task.getStatus() != TaskStatus.INCOMPLETE || task.getDriver().isPresent()
-                || task.getEventTime().isPresent()) {
-            throw new CommandException(MESSAGE_ALREADY_ASSIGNED);
+        // check if the task is already assigned
+        if (task.getStatus().equals(TaskStatus.COMPLETED)) {
+            throw new CommandException(MESSAGE_ALREADY_COMPLETED);
         }
 
+        boolean isAlreadyAssigned = task.getStatus() != TaskStatus.INCOMPLETE || task.getDriver().isPresent()
+                || task.getEventTime().isPresent();
```
Comment (junnbang) [inline src/main/java/seedu/address/logic/commands/AssignCommand.java]: can use the isAssigned() in Task class

File: src/main/java/seedu/address/logic/commands/SuggestCommand.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        Task task = model.getTask(taskId);
+        // check whether the task is scheduled for today
+        if (!task.getDate().equals(GlobalClock.dateToday())) {
+            throw new CommandException(Messages.MESSAGE_NOT_TODAY);
+        }
+
+
+        // check whether the task is already scheduled
+        if (task.getStatus() != TaskStatus.INCOMPLETE || task.getDriver().isPresent()
+                || task.getEventTime().isPresent()) {
```
Comment (junnbang) [inline src/main/java/seedu/address/logic/commands/SuggestCommand.java]: same as above, can use isAssigned() in task class

File: src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        List<Task> taskList = model.getTaskManager().getList();
+
+        return taskList.stream()
+                // find all the task whose customer is the same as the requested task
+                .filter(t -> t.getCustomer().equals(task.getCustomer()))
+
+                // get a distinct stream of drivers of the above tasks
+                .map(Task::getDriver)
+                .flatMap(Optional::stream) // get the list of drivers
+                .distinct()
```
Comment (junnbang) [inline src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java]: this tasklist may include incomplete tasks that is assigned with that customer, so the driver may be empty

File: src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     * he can take up the task.
+     *
+     * @return an Optional of the pair if a suitable driver can be found; otherwise, an empty Optional
+     */
+    private Optional<Candidate> prioritizeSameCustomer() {
+        List<Task> taskList = model.getTaskManager().getList();
+
+        return taskList.stream()
+                // find all the task whose customer is the same as the requested task
+                .filter(t -> t.getCustomer().equals(task.getCustomer()))
```
Comment (junnbang) [inline src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java]: each part of the streams with comments, maybe can abstract out into methods

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 5]
```diff
@@ -31,12 +33,16 @@
 
     public static final String MESSAGE_INVALID_INDEX = "Index is not a non-zero unsigned integer.";
 
+    public static final int MINUTES_IN_AN_HOUR = 60;
+    public static final String HHMM_REGEX = "(0?[1-9]|1[0-2]):[0-5][0-9]";
```
Comment (junnbang) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: how about 00:50?

**Review by junnbang — COMMENTED** [review]
looks fine, just some comments

File: src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java (hunk 3 above)
Comment (tysng) [reply src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java]: Empty driver is taken care of by the flatmap

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 5 above)
Comment (tysng) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: Nice catch!

File: src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java (hunk 4 above)
Comment (tysng) [reply src/main/java/seedu/address/logic/optimizer/ScheduleOptimizer.java]: Very hard to do so without breaking the functional flow of it, save for future maybe.

File: src/main/java/seedu/address/logic/commands/SuggestCommand.java (hunk 2 above)
Comment (tysng) [reply src/main/java/seedu/address/logic/commands/SuggestCommand.java]: `isAssigned` only checks for driver presence, while this is a broader check

**Pull request merged**